Models as classes

jackdobinson opened the pull request:
I spent a few hours with python's regex module and I've ended up with something that could be useful to others.

I needed to have a way to grab model parameter information (model id, parameter name, parameter value, is it fixed or not, etc.) so I can tell what a retrieval is actually doing. The simplest way was to use regular expressions to move all of the code pertaining to models into "ModelClass.py", make each model a class, then add a few extra methods and attributes to enable the model to know where in the state vector its parameters have been stored.

This has had a few effects:

1) The methods: Variables_0::read_apr, Variables_0::calc_NXVAR, ForwardModel_0::subprofretg, and ForwardModel_0::subspecret are now substantially shorter (a chunk of their code is now in "ModelClass.py")

2) "Models.py" now loads the "ModelClass.py" module and stores all the "concrete" model classes in a tuple for easy access by other modules.

3) Looping through VARIDENT in ForwardModel_0 is no longer needed as Variables_0 reads VARIDENT once and creates an array of the models. That array is then used in ForwardModel_0.

4) Adding models is a bit easier, you can just modify a single file and there are subclasses of models to make it a bit easier to tell a user what a model is supposed to do.

5) I've added some default class attributes 'name', and 'description' that I intend to use to make the model classes describe themselves a bit more to make it easier to understand what they do.

6) I've added methods to Variables_0 to enable getting models, their parameters, and plotting their parameters a bit easier. There is an instance property 'parameter_slices' (that has a default implementation on the ModelBase class) that enables you to tell a model what its parameters are called, and where they are stored in 'its section' of the state vector. I've added a proper implementation for model ids 32, 45, 47, and 444 as those are the ones I am using at the moment.

Variables_0 can now make plots like:

![model_parameters](https://github.com/user-attachments/assets/7d78d91f-fed4-4e32-947a-6b844cfd99ad)

juanaldayparejo reviewed (commented):
Hi Jack,

Thanks for doing this. It looks like a more organised way of setting up all the models in a single file, rather than needing to edit several files to include a new model.

However, it represents a substantial change in how the code is structured and I think the documentation should be properly updated so that other users can implement their own models (including me), in principle without any help, just following the documentation.

Before I approve the changes, I was wondering whether you could update the part of the documentation regarding the model parameterisations (https://archnemesis.readthedocs.io/en/latest/documentation/model_parameterisations.html#). I believe it should be the part related to "Custom Models", where we indicate how new models can be implemented. Ideally, we would specify there which files should be changed, and what are the characteristics that each model class should have.

Potentially we could include in Models.py an empty model that users can look at and fill based on their needs? I think that might be useful?

Thanks!
Juan

jackdobinson commented:
Hi Juan,

I have updated the "custom model" documentation to reflect the changes. I have included a walkthrough of creating  a custom model class. Hopefully it will be easy to follow, I am happy to update it if anything is missing/could be better.

I have also split up the models into a 3 different files: Models/AtmosphericModels.py, Models/NonAtmosphericModels.py, Models/SpectralModels.py. These models actually work a little bit differently and it let me tailor the base classes and template classes to minimise the amount of code needed to write a new model.

I've added template class to each file to provide a class someone can copy-paste and modify for their own purposes.

Thanks for having a look.

J

jackdobinson commented:
Hi Juan,

I had some extra thoughts about increasing the new-user friendliness (and I forgot to remove some example code). I've made all the changes I am going to now, so this PR is no longer a moving target.

Thanks,
J

jackdobinson commented:
I found some errors that the tests didn't pick up so I've fixed those :-D

jackdobinson commented:
The *.apr file I was using for testing had comments in it, which was failing to parse. I've swapped the parsing to a method that ignores excess information on each line. Therefore, comments should work again. Also I've added more error messages and warnings to the reading of the *.apr file. Hopefully this will make it easier for the user to diagnose any problems with the *.apr file before a retrieval is run.
